_posts/2023-08-16-debugging-android-devices.md

```diff
@@ -0,0 +1,277 @@
+
+ Of these conditions, the most common cause is when the input dispatching has timed out, which in turn is generally caused by network or file io being done on the main thread. Both of these kinds of work should *always* be done on background threads.
+
+ Please refer to [Android's official documentation](https://developer.android.com/topic/performance/vitals/anr) for more details.
+
+### Java Exceptions
+
+This is the most common type of crash that most Android developers are accustomed to in Android apps and services. In Java, and like many other languages, an exception is an event that occurs during the execution of a program, disrupting the normal flow of instructions. It indicates that something unexpected or erroneous has happened, preventing the program from continuing its normal execution. If you're looking for more information about reading and understanding Java based exceptions, I highly recommend reading [Twilio's excellent blog post](https://www.twilio.com/blog/how-to-read-and-understand-a-java-stacktrace) that goes into great depth about them.
+
+### WTFs
+
+[What Terrible Failures](https://developer.android.com/reference/android/util/Log.html#wtf(java.lang.String,%20java.lang.String)) according to Google should be used to:
+
+> Report a condition that should never happen. The error will always be logged at level ASSERT with the call stack. Depending on system configuration, a report may be added to the DropBoxManager and/or the process may be terminated immediately with an error dialog.
+
+These are logged in Android apps and services throughout the system to record things that as the documentation says, should never happen but we all know that things that should never happen on our systems do still happen, and that's the perfect time to trigger a WTF.
+
+### SeLinux Violations
+
+Android's SELinux (Security-Enhanced Linux) is an implementation of the SELinux framework specifically created by the National Security Agency (NSA). SELinux is used on Android devices to enhance security by enforcing mandatory access controls and ensuring that processes and applications operate within defined security contexts. Android's implementation of SELinux is designed to provide a robust security mechanism to mitigate risks associated with malware, unauthorized access, and other security threats. When incorrectly configured, violations will be logged to logcat via `auditd`. Violations can lead to crashes when the resources attempting to be accessed are protected. One important detail to remember is that in engineering builds, SSLinux is configured to be permissive and will only log a warning but will still allow access to the resource being requested. This can make it easy to miss violations during development. 
+
+You can find more details here in the [official documentation](https://source.android.com/docs/security/features/selinux/).
+
+## Diagnostic Tools
```

Comment:
My main critique is probably that the first half outlining the tools is pretty thorough, but the second half is more of "here's a tool and here's a link to more documentation", and I would've liked to see maybe a few more instances of "I've used this tool in this way to diagnose X specific issue in the past", or "Here's a way to use the tool to diagnose this issue in the field/locally", rather than just explaining the tool. That's a bit tough though.